basic/product-marketplace/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+                                                                zip_code : zip_code
+                                                               } ;   
+
+            let buyer_nft_id = self.buyer_nft_id_by_product_id.get(&product_id).unwrap(); 
+            let buyer_badge = self.buyer_minter_badge_vault.authorize(|auth| {
+                return self.buyer_badge_def.mint_nft(*buyer_nft_id, BuyerNftData {
+                     product_id : product_id,
+                     fees : self.buy_fees,
+                     has_been_sent_by_seller : false 
+                }, auth)
+            }); 
+
+           
+            // set buyer address
+            let seller_nft_id : u128 = *self.seller_nft_id_by_product_id.get(&product_id).unwrap();
+            let mut seller_nft_data  : SellerNftData = self.seller_product_badge_def.get_nft_data(seller_nft_id);  
+            
+            seller_nft_data.buyer_address = buyer_address; 
+            seller_nft_data.product_has_been_purchase = true; 
+            self.seller_product_minter_badge_vault.authorize(|a|  {
+                self.seller_product_badge_def.update_nft_data(seller_nft_id, seller_nft_data,a); 
+            }); 
+
+             // take fees 
+            self.fees_vault.put(payment.take(self.buy_fees));
+            // hold payments
+            let vault = self.payment_by_buyer_nft_id.entry(*buyer_nft_id).or_insert(Vault::new(self.token_type));
+            vault.put(payment.take((*product).price));
+
+            return (buyer_badge, payment);
+        }
+
+        #[auth(seller_product_badge_def)]
+        pub fn collect_postal_stamp(&mut self) -> PostalAddress
+        {
+            let id = auth.get_nft_id(); 
+            let mut seller_nft_data  : SellerNftData = self.seller_product_badge_def.get_nft_data(id); 
+            assert!(!seller_nft_data.postal_stamp_collected, "postale stamp has already collected"); 
+            assert!(seller_nft_data.product_has_been_purchase , 
+                    "product must be purchased");
+                    
+            let buyer_postal_address :PostalAddress = PostalAddress {
+                 street : seller_nft_data.buyer_address.street.clone(),
+                 city : seller_nft_data.buyer_address.city.clone(),
+                 zip_code : seller_nft_data.buyer_address.zip_code.clone()
+            };
+
+            seller_nft_data.postal_stamp_collected = true; 
+            self.seller_product_minter_badge_vault.authorize(|a|  {
+                self.seller_product_badge_def.update_nft_data(id, seller_nft_data, a); 
+            }); 
+
+            return buyer_postal_address;
+        }
+    
+        #[auth(seller_product_badge_def)]
+        pub fn send_product(&mut self)
+        {
+            let seller_nft_id = auth.get_nft_id(); 
+            let mut seller_nft_data  : SellerNftData = self.seller_product_badge_def.get_nft_data(seller_nft_id); 
+            assert!(seller_nft_data.product_has_been_purchase , 
```

Comment:
You might also want to assert that `seller_nft_data.postal_stamp_collected == true` ?

Reply:
I guess the seller can send the product without collecting the postage stamp.

Reply:
I agree. Just wanted to verify that it was not a mistake